### ISSUE TITLE
Ground truth preprocessing

### ISSUE THREAD

**Issue by sampepose** [lmb-freiburg/demon#12]
Hello,

I am trying to finetune your iterative network with my own dataset.

I understand the ground truth must be:
- image pair [bs, 6, 192, 256] in range [-0.5, 0.5]
- inverse depth [bs, 1, 48, 64] in range [0.0, inf]
- normals [bs, 3, 48, 64] in range [-1.0, 1.0]
- optical flow [bs, 2, 48, 64] in range [-1.0, 1.0] bc x/y components scaled by image width/height
- rotation [bs, 3] where norm is the angle
- translation [bs, 3] which is a unit vector

However, at the start of training, I'm comparing the ranges of the ground truth to predictions from the iterative net:
> [name][min][max]

> [gt depth][0.0015259022][0.25714538]
> [pred depth][-4.7255404e+09][4.8389247e+10]

> [gt flow][-0.33519724][0.31490907]
> [pred flow][-1.160358][0.12831117]

> [gt normals][-1][0.99998885]
> [pred normals][-3502147.5][2898221.5]

The ranges look reasonable for flow, however the predictions for depth and normals are much larger than the gt. I am generating the gt normals using the depth_to_normals function from lmbspecialops with my gt inverse depth.

Am I missing something with my preprocessing? For reference, I'm using the virtual kitti dataset.

Thank you for your time,
Sam

**Comment by benjaminum** [comment]
The ranges for the ground truth data are basically correct. inf for inverse depth should be avoided and the optical flow can be larger or smaller than 1.0 or -1.0 but in practice this does rarely happens.

I assume that the values of the bootstrap net look reasonable.

Is this the first iterations of the iterative net?
Can you check what the output of the flow_to_depth operation looks like?
 This is in the ```depthmotion_block_demon_original```  function in ```blocks_original.py``` part? 
https://github.com/lmb-freiburg/demon/blob/72625d2f0b3f30757aa48b29892d02739805ec87/python/depthmotionnet/blocks_original.py#L343-L351

**Comment by sampepose** [comment]
Yes, the results of the bootstrap net look reasonable. Here are histograms of ground truth and output of bootstrap net:

| gt  | output of bootstrap |
| ------------- | ------------- |
| <img width="246" alt="screenshot 2017-08-14 10 10 29" src="https://user-images.githubusercontent.com/1855931/29275441-17e38712-80d9-11e7-86d7-8cb914359f81.png">  | <img width="256" alt="screenshot 2017-08-14 10 10 37" src="https://user-images.githubusercontent.com/1855931/29275444-1afbdc42-80d9-11e7-974f-f1bfdff1e715.png"> |
| <img width="251" alt="screenshot 2017-08-14 10 11 39" src="https://user-images.githubusercontent.com/1855931/29275477-3bc13dfa-80d9-11e7-96f1-6e760ffb2d40.png">  | <img width="256" alt="screenshot 2017-08-14 10 11 35" src="https://user-images.githubusercontent.com/1855931/29275487-43c4e236-80d9-11e7-837d-35160e6008b4.png">  |

This is the first iteration of the iterative net.

Here's the min/max/avg of the output of flow_to_depth for a few training iterations.
[0][179.47673][0.16061339]
**[0][1.0737418e+09][3646125]**
[0][903.28778][0.2184152]
[0][209.7283][0.38293144]
[0][1789.4653][12.30909]
[0][180.58949][1.5996453]

I'm not sure why the second iteration there has a crazy high max value.

The output of the iterative net visualized:

| gt  | output of iterative |
| ------------- | ------------- |
| <img width="248" alt="screenshot 2017-08-14 10 28 23" src="https://user-images.githubusercontent.com/1855931/29276079-64b175de-80db-11e7-8d86-75a88406605f.png"> | <img width="256" alt="screenshot 2017-08-14 10 28 40" src="https://user-images.githubusercontent.com/1855931/29276085-691cad3c-80db-11e7-947b-9dd478a7b050.png"> |
|<img width="249" alt="screenshot 2017-08-14 10 28 29" src="https://user-images.githubusercontent.com/1855931/29276090-6cda97e0-80db-11e7-8f60-e8b1b1c53be6.png"> | <img width="251" alt="screenshot 2017-08-14 10 28 44" src="https://user-images.githubusercontent.com/1855931/29276092-6f951f0a-80db-11e7-9335-f0e3d6c0bbe7.png"> |


Depth predictions are in the range of [-1E8, 1E8] and normals in the range [-4E4, 8E4]

**Comment by sampepose** [comment]
I printed the inputs to flow_to_depth as well, and they look reasonable.

[name][min][max][avg]

[translation][-0.96468806][0.37828878][-0.27458572]
[rotation][-0.021255214][0.012888308][-0.0051129474]
[flow][-0.06706503][0.13209946][0.0039939629]
**[depth from flow][0][33554432][1267849]**

[translation][-0.99621952][0.37635651][-0.31681082]
[rotation][-0.020021241][0.043232266][-0.0013534732]
[flow][-0.043860085][0.082544379][0.0022001718]
**[depth from flow][0][1.3421773e+08][5803308.5]**

**Comment by sampepose** [comment]
The structure of the given flow looks pretty reasonable too.
<img width="1008" alt="screenshot 2017-08-14 18 22 01" src="https://user-images.githubusercontent.com/1855931/29294551-e8c38d02-811d-11e7-9878-889a08df36ef.png">

The ranges are roughly the same (ground truth vs iterative net prediction):
<img width="563" alt="screenshot 2017-08-14 18 27 23" src="https://user-images.githubusercontent.com/1855931/29294621-3bebecd6-811e-11e7-8b78-3788e598f2ed.png">

**Comment by sampepose** [comment]
I verified the op looks fine for your given example:
![iter0](https://user-images.githubusercontent.com/1855931/29296217-e604755a-8126-11e7-9b0c-dc9510875d3f.png)

However, it doesn't look fine _even with my ground truth flow, rotation, and translation!_
![image](https://user-images.githubusercontent.com/1855931/29296248-11474f94-8127-11e7-9220-e747463f2c80.png)
![image](https://user-images.githubusercontent.com/1855931/29296255-1a770442-8127-11e7-996b-a71040f42071.png)
![image](https://user-images.githubusercontent.com/1855931/29296261-2513fd74-8127-11e7-84b6-a4a5b2eedcda.png)

- Ground truth flow is normalized by dividing x,y components by image width and height, respectively.
- Ground truth translation is divided by the norm to make it a unit vector
- Ground truth rotation is given as a rotation matrix, so I convert this to a rotation axis with a magnitude equal to the rotation angle.
- Intrinsics are left alone to your defaults

**Comment by sampepose** [comment]
One issue is the flow_to_depth op always expects intrinsics to be a (N, 4) tensor. This works fine with a batch size of 1 like the example. However, has unexpected results with larger batch sizes. The code should either be changed to use a single intrinsics matrix for all images in the batch (makes sense) or the validation should check for the correct matrix rank.

Nonetheless, after fixing and recompiling this error locally (and verifying that I am getting valid intrinsic and fundamental matrices for every image in the batch), the depth maps still look like in my last comment.

**Comment by benjaminum** [comment]
> One issue is the flow_to_depth op always expects intrinsics to be a (N, 4) tensor. See here: https://github.com/lmb-freiburg/lmbspecialops/blob/master/src/flowtodepth.cc#L361
> 
> This works fine with a batch size of 1 like the example. However, has unexpected results with larger batch sizes. The code should either be changed to use a single intrinsics matrix for all images in the batch (makes sense) or the validation should check for the correct matrix rank.

Thanks for pointing this out. The documentation of the op is misleading. I will adapt the documentation and add checks. We want to allow different intrinsics for each batch item to be more flexible.

Your comments definitely help to debug this. From your last comment it seems that there is something seriously wrong with the op. 
Would it be possible to export the inputs and outputs as numpy array (.npy) and attach them here? I will try to have a look as soon as possible, but I cannot give guarantees how quickly this can be resolved.

If you need a workaround fast, you can replace the output of the flow_to_depth op with the previously predicted depth during finetuning.

**Comment by sampepose** [comment]
Thanks, Benjamin. I've attached a zip with my ground truth as .npy.

[gt.zip](https://github.com/lmb-freiburg/demon/files/1228735/gt.zip)

The main issue is the range produced by flow_to_depth is way too large to produce proper visualization. When normalized to [0, 255], almost all of the values are 0. Here's a box plot of the values:
![boxplot](https://user-images.githubusercontent.com/1855931/29372373-eb395efc-8278-11e7-8acd-11e500e51182.png)

The median is way down at 0.X but there are a lot of values outside of the whiskers. The top whisker is at Q3 + 1.5*IQR.

**Comment by benjaminum** [comment]
We have now added a ```flow_to_depth2``` op with bugfixes.

The data in ```gt.zip``` however, still looks strange. Does the scale of the depth value correspond to the unit translation?

**Comment by sampepose** [comment]
Thanks, Benjamin.

The scale is definitely off. Given a non-unit translation, what is the
proper way of fixing the scale of the ground truth depth/normals when
making it a unit translation?

SP

**Comment by benjaminum** [comment]
You simply scale the depth (not inverse depth) with the same scalar factor as the translation.

**Comment by sampepose** [comment]
Here's ground truth with the depth properly scaled. The depth in this .zip is inverse depth.

[gt.zip](https://github.com/lmb-freiburg/demon/files/1239911/gt.zip)
